File: Python/Sorting/Insertion_Sort.py
#------------Insertion Sort complexity----------------
# Best Case - O(n)             |--> If the array is nearly sorted
# Average Case - O(n^2)
# Worst Case - O(n^2)

# Space Complexity - O(1)

def insertion_Sort(array):
    # Traverse through 1 to len(arr) 
    for i in range(1, len(array)): 
  
        #storing the key value to find the position in array
        key = array[i] 
  
        # Move elements of arr[0..i-1], that are 
        # greater than key, to one position ahead 
        # of their current position 
        j = i-1
        while j >= 0 and key < array[j] : 
                array[j + 1] = array[j] 
                j -= 1
        array[j + 1] = key
        #To see the partial sorted array after every pass remove the comment of below code
        #print(array)
    
    return array

print(insertion_Sort([6,2,4,1,8,5]))

# Step 1: [ 2, 6, 4, 1, 8, 5 ]     currentVal=6
# Step 2: [ 2, 4, 6, 1, 8, 5 ]     currentVal=4
# Step 3: [ 1, 2, 4, 6, 8, 5 ]     currentVal=1
# Step 4: [ 1, 2, 4, 6, 8, 5 ]     currentVal=8
# Step 5: [ 1, 2, 4, 5, 6, 8 ]     currentVal=5
# Step 6: [ 1, 2, 4, 5, 6, 8 ]

File: C++/sorting/insertion_sort.cpp
 //------------Insertion Sort complexity----------------
// Best Case - O(n)             |--> If the array is nearly sorted
// Average Case - O(n^2)
// Worst Case - O(n^2)
// Space Complexity - O(1)
#include<iostream>
using namespace std;

void insertion_sort(int arr[], int array_size)
{
    int current_value;
     // Traverse through 1 to length of Array
    for(int i=1;i<array_size;i++)
    {
        int j;
        //storing the key value to find the position in array
        current_value=arr[i];
        // Move elements of arr[0..i-1], that are 
        // greater than key, to one position ahead 
        // of their current position 
        for(j=i-1; j>=0 && arr[j] > current_value; j--)
        {
            arr[j+1] = arr[j];
        }
        arr[j+1] = current_value;
        //To see the partial sorted array after every pass remove the comment of below code
        
        
    }
    
}

void print_array(int arr[], int array_size) // to print the array when called
{
    for(int i=0;i<array_size;i++)
    {
        cout<<arr[i];
    }
}

int main()
{
    int arr[] = {-2, 45, 0, 11, -9};
    int size = sizeof(arr) / sizeof(arr[0]);    //to find the length of the array.
    insertion_sort(arr, size);                     //function is called and the array is sorted.
    cout << "Sorted Array in Ascending Order:\n";
    print_array(arr, size);                     //function is called to print the sorted array.
    return 0;
}